src/awkward/forms/form.py

```diff
@@ -452,19 +462,29 @@ def length_zero_array(
             simplify=False,
         )
 
+    def _smallest_zero_buffer_lengths(self) -> Iterator[ShapeItem]:
+        raise NotImplementedError
+
     def length_one_array(self, *, backend=numpy_backend, highlevel=True, behavior=None):
-        # A length-1 array will need at least N bytes, where N is the largest dtype (e.g. 256 bit complex)
-        # Similarly, a length-1 array will need no more than 2*N bytes, as all contents need at most two
-        # index-types e.g. `ListOffsetArray.offsets` for their various index metadata. Therefore, we
-        # create a buffer of this length (2N) and instruct all contents to use it (via `buffer_key=""`).
-        # At the same time, with all index-like metadata set to 0, the list types will have zero lengths
-        # whilst unions, indexed, and option types will contain a single value.
+        # The naive implementation of a length-1 array requires that we have a sufficiently
+        # large buffer to be able to build _any_ subtree.
+        def max_prefer_unknown(this: ShapeItem, that: ShapeItem) -> ShapeItem:
+            if this is unknown_length:
+                return this
+            if that is unknown_length:
+                return that
+            return max(this, that)
+
+        buffer_length = reduce(max_prefer_unknown, self._smallest_zero_buffer_lengths())
+        if buffer_length is unknown_length:
+            raise NotImplementedError(
+                "cannot create length_zero_array from a form containing unknown shape items"
+            )
```

Comment:
This implementation of `_smallest_zero_buffer_lengths` is designed to avoid the call to `max` until the top level. This is a defensive approach; I'm not exactly clear on how this will tie in with typetracer / unknown shapes over time, so for now it's easy to error in a single place.

Reply:
That's a good way to write the `max` function in a typetracer-agnostic way. It's true that the max of `x` and unknown is unknown.

The intended use of this function is to turn a typetracer layout into a length-1 non-typetracer array. The typetracer might have `RegularArray.size is unknown_length`. That's the only way that an actual unknown length could come into this (the outermost length of the typetracer layout is usually unknown, but we'll be replacing that unknown length with 1.)

What kind of array would you want in that situation, @lgray? If an internal `RegularArray.size is unknown_length`, would you want an array with `RegularArray.size == 0` or maybe `1`? For the application, it might not matter, or you might want to pass a default down the recursion.

Reply:
For my particular use case I'm generally adding a batching dimension for ML evaluation, which is unknown. It will always be the outermost dimension.

Reply:
But in general, length one for unknowns makes the most sense to me, given the same of the function and the intent.

Reply:
@jpivarski we now also have `unknown_length` in `known_data` contexts. It's being used for the re-hydration PR. I'm not very comfortable with that, but equally choosing an arbitrary "wrong" length is not the solution there either.

Reply:
Yeah, but suppose that batch should have regular shape `?? * 50 * 123 * ?? * 10`. The outer `??` is unknown because it's a batch—that's fine—but what do we do about any inner `??` that might exist? In a non-typetracer array, this will resolve to a known integer, but it's unknown in the Dask graph/typetracer pass because it was derived from a value of some other array. (Maybe it was ragged data that's getting padded by the longest list, which isn't known until you have concrete arrays.) When you call `length_one_if_typetracer` in this instance, can we insert a `0`, a `1`, or a default you pass down to fill in the inner `??`?

Reply:
Passing a 1 makes the most sense IMO.

Reply:
We can't let an `unknown_length` be used in a rehydrated (non-typetracer) array. But it might be appropriate to replace all of those inner `RegularArray.size is unknown_length` with `0` or `1`. Given what it's used for (rehydrated arrays are supposed to never be touched), it might not be wrong.

Reply:
The main discrepancy with rehydration is that it doesn't track `shape_touched`, so we might depend upon the shape. A common example would be broadcasting `NumpyArray`; these don't need to be touched, but they must have compatible shapes. After rehydration, any arbitrary length we choose won't be guaranteed to agree with the other layout.

Reply:
A length-1 regular dimension can agree with any broadcasting. (Fortunately!)